rust/psibase/src/table.rs

```diff
@@ -208,41 +208,92 @@ impl<Key: ToKey, Record: TableRecord> TableIndex<Key, Record> {
-    /// It does not affect the iterator
     pub fn get(&self, key: Key) -> Option<Record> {
         let mut key_bytes = self.prefix.clone();
         key.append_key(&mut key_bytes);
 
         kv_get_bytes(self.db_id, &key_bytes).and_then(|v| self.get_value_from_bytes(v))
     }
 
-    /// Set the range of available records for the iterator; useful for viewing a slice of the records
-    ///
-    /// - `from` is inclusive, `to` is not: [from..to)
-    /// - using this after the start of an iteration process (for..loops, next(), next_back() etc)
-    /// can mess the iteration sequence order, since it resets the front and the back cursors
-    pub fn range(&mut self, from: Key, to: Key) -> &mut Self {
+    /// Creates an iterator for visiting all the table key-values, in sorted order.
+    pub fn iter(&self) -> TableIter<'_, Key, Record> {
+        TableIter {
+            table_index: self,
+            front_key: None,
+            back_key: None,
+            is_end: false,
+        }
+    }
+
+    /// Constructs a double-ended iterator over a sub-range of elements in the table.
+    /// The simplest way is to use the range syntax `min..max`, thus `range(min..max)` will
+    /// yield elements from min (inclusive) to max (exclusive).
+    /// The range may also be entered as `(Bound<T>, Bound<T>)`, so for example
+    /// `range((Excluded(4), Included(10)))` will yield a left-exclusive, right-inclusive
+    /// range from 4 to 10.
+    pub fn range(&self, range: impl RangeBounds<Key>) -> TableIter<'_, Key, Record> {
         let mut front_key = self.prefix.clone();
-        from.append_key(&mut front_key);
-        self.front_key = Some(RawKey::new(front_key));
+        let front_key = match range.start_bound() {
+            Bound::Included(k) => {
+                println!("included fc");
+                k.append_key(&mut front_key);
+                Some(front_key)
+            }
+            Bound::Excluded(k) => {
+                println!("excluded fc");
+                k.append_key(&mut front_key);
+                front_key.push(0);
+                Some(front_key)
+            }
+            Bound::Unbounded => None,
+        }
+        .map(RawKey::new);
+        if front_key.is_some() {
+            println!("range2 fc: {}", to_hex(&front_key.as_ref().unwrap().data));
+        }
 
         let mut back_key = self.prefix.clone();
-        to.append_key(&mut back_key);
-        self.back_key = Some(RawKey::new(back_key));
+        let back_key = match range.end_bound() {
+            Bound::Included(k) => {
+                println!("excluded bc");
+                k.append_key(&mut back_key);
+                // since the kv_less_than is used, increment last byte to be included
+                let mut last_byte = back_key.pop().unwrap();
+                if last_byte == u8::MAX {
+                    back_key.push(last_byte);
+                    back_key.push(0);
+                } else {
+                    last_byte += 1;
```

Comment:
Let's say `k` is `..., 7`. This sets back_key to `..., 8`. This includes `..., 7, x, ...` in the range even though it is after `..., 7`.

Reply:
Nice catch, fixed!